hyperparameter_tuning.py

```diff
@@ -0,0 +1,155 @@
+        else:
+            train_inputs = kwargs
+        X_valid, y_valid = kwargs[self.valid_input_keys[0]], kwargs[self.valid_input_keys[1]]
+
+        results = []
+        for i, param_set in enumerate(self.param_space):
+            try:
+                transformer = self.TransformerClass(**param_set)
+                transformer.fit(**train_inputs)
+            except Exception:
+                continue
+            y_pred_valid = transformer.transform(X_valid)
+            y_pred_valid_value = list(y_pred_valid.values())[0]
+            run_score = self.score_func(y_valid, y_pred_valid_value)
+            results.append((run_score, param_set))
+
+            del y_pred_valid, transformer
+            gc.collect()
+
+            for callback in self.callbacks:
+                callback.on_run_end(score=run_score, params=param_set)
+
+        assert len(results) > 0, 'All random search runs failed, check your parameter space'
+        results_sorted = sorted(results, key=lambda x: x[0])
+
+        if self.maximize:
+            best_score, best_param_set = results_sorted[-1]
+        else:
+            best_score, best_param_set = results_sorted[0]
+
+        for callback in self.callbacks:
+            callback.on_search_end(results=results)
+
+        self.best_transformer = self.TransformerClass(**best_param_set)
+        self.best_transformer.fit(**train_inputs)
+        return self
+
+    def transform(self, **kwargs):
+        return self.best_transformer.transform(**kwargs)
+
+    def save(self, filepath):
+        self.best_transformer.save(filepath)
+
+    def load(self, filepath):
+        self.best_transformer.load(filepath)
+        return self
+
+
+def create_param_space(params, n_runs):
+    seed = np.random.randint(1000)
+    param_space = []
+    for i in range(n_runs):
+        set_seed(seed + i)
+        param_choice = {}
+        for param, value in params.items():
+            if isinstance(value, list):
+                if len(value) == 2:
+                    mode = 'choice'
+                    param_choice[param] = sample_param_space(value, mode)
+                else:
+                    mode = value[-1]
+                    param_choice[param] = sample_param_space(value[:-1], mode)
+            else:
+                param_choice[param] = value
+        param_space.append(param_choice)
+    return param_space
+
+
+def sample_param_space(value_range, mode):
+    if mode == 'list':
+        value = np.random.choice(value_range)
+    else:
+        range_min, range_max = value_range
+        if mode == 'choice':
+            value = np.random.choice(range(range_min, range_max, 1))
+        elif mode == 'uniform':
+            value = np.random.uniform(low=range_min, high=range_max)
+        elif mode == 'log-uniform':
+            value = np.exp(np.random.uniform(low=np.log(range_min), high=np.log(range_max)))
+        else:
+            raise NotImplementedError
+    return value
+
+
+class GridSearchCallback:
```

Comment:
All callbacks should be placed in single Python file. We should discuss where to place them (in the steppy-toolkit).